File: gpuwm/static/lambert.py
"""Lambert conformal projection grids with WPS conventions.

Pure NumPy float64, setup-time only (no GPU).  The projection math is
transcribed from WRF v4.6.1 ``share/module_llxy.F`` (``lc_cone``:1138,
``set_lc``:1097, ``ijll_lc``:1174, ``llij_lc``:1250 -- the same code WPS
geogrid links); the derived fields follow the ARW tech note sec 2 / Snyder
with WPS geogrid conventions.  Where a convention could be read two ways the
bundle geo_em files arbitrate (tests/test_lambert.py gates every field on
all four 1974 domains against geogrid v4.6.0 output), and since the
worldwide lane the pinned-source oracle fixture arbitrates both
hemispheres (tests/test_projection_oracle.py, built by
tools/llxy_wrf461_oracle from the pristine pinned source):

- Grid registration: 1-based mass point (i, j) sits at projection
  coordinate (i, j); U at (i - 0.5, j); V at (i, j - 0.5); corner (C) at
  (i - 0.5, j - 0.5).  The reference point (ref_lat, ref_lon) sits at
  (e_we/2, e_sn/2), WPS's default ref_x/ref_y (oracle: XLAT/XLONG to
  ~3e-5 deg, the float32-storage bound).
- Map factor  m(lat) = cos(tl1)/cos(lat)
  * (tan(colat/2) / tan(colat1/2))**cone  -- equals 1 at both true
  latitudes (oracle: MAPFAC_* to ~3e-7 relative).  This is the ARW
  tech-note form referenced to truelat1; geogrid's get_map_factor uses
  the mathematically identical form referenced to truelat2, so the
  binary64 oracle gate for this one quantity is a small relative bound
  rather than a ULP count (recorded in tests/test_projection_oracle.py).
- Coriolis  F = 2*OMEGA_E*sin(lat), E = 2*OMEGA_E*cos(lat)  with WPS's
  ``OMEGA_E = 7.292e-5`` from constants_module (oracle: ~9e-7 relative;
  WRF's ``EOMEG = 7.2921e-5`` does *not* reproduce geogrid's F/E).
- Wind rotation  alpha = cone * (stand_lon - lon):  geo_em
  SINALPHA == sin(alpha), COSALPHA == cos(alpha), transcribed from WPS
  geogrid get_rotang (process_tile_module.F:1920), which applies NO
  hemisphere factor -- the cone from lc_cone is positive in both
  hemispheres.  Divergence record: before the worldwide lane this
  module multiplied alpha by ``hemi``, an unreachable SH guess (the
  wizard refused the southern hemisphere); on every reachable NH grid
  ``hemi`` was exactly 1.0, so all shipped fields are byte-identical.
- Nest registration (namelist parent_start/ratio arithmetic): nest mass
  coordinate x maps to parent mass coordinate
  ``(i_parent_start - 0.5) + (x - 0.5)/ratio`` -- cell edges align, and for
  odd ratios every centre nest point coincides with a parent point (oracle:
  geo_em.d02/d03/d04 to ~5e-5 deg through the d02->d03->d04 chain).

Mercator and polar stereographic live in :mod:`gpuwm.static.projection`
together with the shared grid machinery; this module re-exports the
namelist/config dispatch helpers for compatibility.
"""
from __future__ import annotations

import numpy as np

from gpuwm.static.projection import (  # noqa: F401  (compat re-exports)
    EARTH_RADIUS_M, OMEGA_E, ProjectedGrid, _DEG_PER_RAD, _RAD_PER_DEG,
    _parse_wps_namelist, _wrap180, grids_from_projection_config,
    grids_from_wps_namelist)


def _lc_cone(truelat1: float, truelat2: float) -> float:
    """Cone constant, transcribed from module_llxy.F lc_cone (:1138)."""
    if abs(truelat1 - truelat2) > 0.1:  # secant
        num = (np.log10(np.cos(truelat1 * _RAD_PER_DEG))
               - np.log10(np.cos(truelat2 * _RAD_PER_DEG)))
        den = (np.log10(np.tan((45.0 - abs(truelat1) / 2.0) * _RAD_PER_DEG))
               - np.log10(np.tan((45.0 - abs(truelat2) / 2.0) * _RAD_PER_DEG)))
        return float(num / den)
    return float(np.sin(abs(truelat1) * _RAD_PER_DEG))  # tangent


class LambertGrid(ProjectedGrid):
    """One WRF domain on a Lambert conformal (secant or tangent) projection.

    Parameters mirror the WPS &geogrid namelist: ``(ref_lat, ref_lon)`` is
    pinned at projection coordinate ``(known_x, known_y)`` which defaults to
    the WPS ref_x/ref_y of ``(e_we/2, e_sn/2)`` -- the mass-grid centre.
    ``e_we``/``e_sn`` are the *staggered* dimensions (mass grid is one
    smaller).  ``dy`` must equal ``dx`` (WPS requirement for Lambert).

    All array methods return float64 arrays in ``(south_north, west_east)``
    order, matching geo_em's ``[j, i]`` layout.
    """

    map_proj = "lambert"

    def _setup(self) -> None:
        # --- set_lc transcription (module_llxy.F:1097) ---
        self.cone = _lc_cone(self.truelat1, self.truelat2)
        self.rebydx = EARTH_RADIUS_M / self.dx
        deltalon1 = float(_wrap180(self.ref_lon - self.stand_lon))
        ctl1r = np.cos(self.truelat1 * _RAD_PER_DEG)
        # radius (grid lengths) from the pole to the known point
        self.rsw = float(
            self.rebydx * ctl1r / self.cone
            * (np.tan((90.0 * self.hemi - self.ref_lat) * _RAD_PER_DEG / 2.0)
               / np.tan((90.0 * self.hemi - self.truelat1)
                        * _RAD_PER_DEG / 2.0)) ** self.cone)
        arg = self.cone * (deltalon1 * _RAD_PER_DEG)
        self.polei = float(self.hemi * self.known_x
                           - self.hemi * self.rsw * np.sin(arg))
        self.polej = float(self.hemi * self.known_y
                           + self.rsw * np.cos(arg))

    # -- core transforms ---------------------------------------------------

    def ij_to_latlon(self, x, y):
        """Projection coordinate (x, y) -> (lat, lon) degrees.

        Transcription of ijll_lc (module_llxy.F:1174), vectorized.
        """
        x = np.asarray(x, dtype=np.float64)
        y = np.asarray(y, dtype=np.float64)
        chi1 = (90.0 - self.hemi * self.truelat1) * _RAD_PER_DEG
        chi2 = (90.0 - self.hemi * self.truelat2) * _RAD_PER_DEG
        xx = self.hemi * x - self.polei
        yy = self.polej - self.hemi * y
        r2 = xx * xx + yy * yy
        r = np.sqrt(r2) / self.rebydx
        lon = self.stand_lon + _DEG_PER_RAD * np.arctan2(self.hemi * xx,
                                                         yy) / self.cone
        lon = np.mod(lon + 360.0, 360.0)
        if chi1 == chi2:  # tangent (exact-equality branch, as in Fortran)
            chi = 2.0 * np.arctan((r / np.tan(chi1)) ** (1.0 / self.cone)
                                  * np.tan(chi1 * 0.5))
        else:             # secant
            chi = 2.0 * np.arctan(
                (r * self.cone / np.sin(chi1)) ** (1.0 / self.cone)
                * np.tan(chi1 * 0.5))
        lat = (90.0 - chi * _DEG_PER_RAD) * self.hemi
        # pole point (r2 == 0): the formulas above already yield
        # lat = 90*hemi, lon = stand_lon in IEEE arithmetic, but mirror the
        # Fortran branch explicitly for exactness.
        lat = np.where(r2 == 0.0, 90.0 * self.hemi, lat)
        lon = np.where(r2 == 0.0, np.mod(self.stand_lon + 360.0, 360.0), lon)
        lon = np.where(lon > 180.0, lon - 360.0, lon)
        lon = np.where(lon < -180.0, lon + 360.0, lon)
        return lat, lon

    def latlon_to_ij(self, lat, lon):
        """(lat, lon) degrees -> projection coordinate (x, y).

        Transcription of llij_lc (module_llxy.F:1250), vectorized.
        """
        lat = np.asarray(lat, dtype=np.float64)
        lon = np.asarray(lon, dtype=np.float64)
        deltalon = _wrap180(lon - self.stand_lon)
        ctl1r = np.cos(self.truelat1 * _RAD_PER_DEG)
        rm = (self.rebydx * ctl1r / self.cone
              * (np.tan((90.0 * self.hemi - lat) * _RAD_PER_DEG / 2.0)
                 / np.tan((90.0 * self.hemi - self.truelat1)
                          * _RAD_PER_DEG / 2.0)) ** self.cone)
        arg = self.cone * (deltalon * _RAD_PER_DEG)
        x = self.polei + self.hemi * rm * np.sin(arg)
        y = self.polej - rm * np.cos(arg)
        return self.hemi * x, self.hemi * y

    # -- derived fields ------------------------------------------------------

    def map_factor(self, lat):
        """Map scale factor m(lat); equals 1 at both true latitudes."""
        lat = np.asarray(lat, dtype=np.float64)
        half_colat = (90.0 * self.hemi - lat) * _RAD_PER_DEG / 2.0
        half_colat1 = (90.0 * self.hemi - self.truelat1) * _RAD_PER_DEG / 2.0
        return (np.cos(self.truelat1 * _RAD_PER_DEG)
                / np.cos(lat * _RAD_PER_DEG)
                * (np.tan(half_colat) / np.tan(half_colat1)) ** self.cone)

    def rotation(self, lon):
        """(SINALPHA, COSALPHA) at longitudes LON.

        alpha = cone * (stand_lon - lon): geogrid get_rotang PROJ_LC
        (no hemisphere factor; cone is positive in both hemispheres).
        Operation order preserves the shipped NH bytes: the retired
        ``hemi *`` prefactor was exactly 1.0 on every reachable grid.
        """
        lon = np.asarray(lon, dtype=np.float64)
        alpha = (self.cone * _RAD_PER_DEG
                 * _wrap180(self.stand_lon - lon))
        return np.sin(alpha), np.cos(alpha)

File: tools/rustwx/crates/static-fields/src/projection/lambert.rs
//! LANE 1.  Lambert conformal (secant/tangent), transcribed from
//! module_llxy.F `lc_cone`:1138, `set_lc`:1097, `ijll_lc`:1174,
//! `llij_lc`:1250 -- see `gpuwm/static/lambert.py` for the arbitration
//! notes (grid registration, map factor referenced to truelat1,
//! OMEGA_E Coriolis, get_rotang with NO hemisphere factor).
//!
//! Operation order matches the Python float64 spelling exactly (left
//! associativity preserved); byte parity is gated by
//! `tests/lane1_goldens.rs` against arrays extracted from the real
//! Python on the harness domains.

use super::npmath::{np_mod, np_pow};
use super::{wrap180, GridSpec, DEG_PER_RAD, RAD_PER_DEG};
use crate::EARTH_RADIUS_M;

/// set_lc outputs.
#[derive(Debug, Clone)]
pub struct LambertState {
    pub cone: f64,
    pub rebydx: f64,
    pub rsw: f64,
    pub polei: f64,
    pub polej: f64,
}

/// Cone constant, transcribed from module_llxy.F lc_cone (:1138).
pub fn lc_cone(truelat1: f64, truelat2: f64) -> f64 {
    if (truelat1 - truelat2).abs() > 0.1 {
        // secant
        let num = (truelat1 * RAD_PER_DEG).cos().log10()
            - (truelat2 * RAD_PER_DEG).cos().log10();
        let den = ((45.0 - truelat1.abs() / 2.0) * RAD_PER_DEG).tan().log10()
            - ((45.0 - truelat2.abs() / 2.0) * RAD_PER_DEG).tan().log10();
        num / den
    } else {
        (truelat1.abs() * RAD_PER_DEG).sin() // tangent
    }
}

/// set_lc transcription (module_llxy.F:1097).
pub fn setup(spec: &GridSpec, hemi: f64) -> LambertState {
    let cone = lc_cone(spec.truelat1, spec.truelat2);
    let rebydx = EARTH_RADIUS_M / spec.dx;
    let deltalon1 = wrap180(spec.ref_lon - spec.stand_lon);
    let ctl1r = (spec.truelat1 * RAD_PER_DEG).cos();
    let rsw = rebydx * ctl1r / cone
        * np_pow(
            ((90.0 * hemi - spec.ref_lat) * RAD_PER_DEG / 2.0).tan()
                / ((90.0 * hemi - spec.truelat1) * RAD_PER_DEG / 2.0).tan(),
            cone,
        );
    let arg = cone * (deltalon1 * RAD_PER_DEG);
    let polei = hemi * spec.known_x - hemi * rsw * arg.sin();
    let polej = hemi * spec.known_y + rsw * arg.cos();
    LambertState { cone, rebydx, rsw, polei, polej }
}

/// ijll_lc transcription (module_llxy.F:1174), one point.
pub fn ij_to_latlon(
    state: &LambertState,
    spec: &GridSpec,
    hemi: f64,
    x: f64,
    y: f64,
) -> (f64, f64) {
    let chi1 = (90.0 - hemi * spec.truelat1) * RAD_PER_DEG;
    let chi2 = (90.0 - hemi * spec.truelat2) * RAD_PER_DEG;
    let xx = hemi * x - state.polei;
    let yy = state.polej - hemi * y;
    let r2 = xx * xx + yy * yy;
    let r = r2.sqrt() / state.rebydx;
    let mut lon =
        spec.stand_lon + DEG_PER_RAD * (hemi * xx).atan2(yy) / state.cone;
    lon = np_mod(lon + 360.0, 360.0);
    let chi = if chi1 == chi2 {
        // tangent (exact-equality branch, as in Fortran)
        2.0 * (np_pow(r / chi1.tan(), 1.0 / state.cone) * (chi1 * 0.5).tan())
            .atan()
    } else {
        // secant
        2.0 * (np_pow(r * state.cone / chi1.sin(), 1.0 / state.cone)
            * (chi1 * 0.5).tan())
        .atan()
    };
    let mut lat = (90.0 - chi * DEG_PER_RAD) * hemi;
    // pole point (r2 == 0): mirror the Fortran branch explicitly.
    if r2 == 0.0 {
        lat = 90.0 * hemi;
        lon = np_mod(spec.stand_lon + 360.0, 360.0);
    }
    if lon > 180.0 {
        lon -= 360.0;
    }
    if lon < -180.0 {
        lon += 360.0;
    }
    (lat, lon)
}

/// llij_lc transcription (module_llxy.F:1250), one point.
pub fn latlon_to_ij(
    state: &LambertState,
    spec: &GridSpec,
    hemi: f64,
    lat: f64,
    lon: f64,
) -> (f64, f64) {
    let deltalon = wrap180(lon - spec.stand_lon);
    let ctl1r = (spec.truelat1 * RAD_PER_DEG).cos();
    let rm = state.rebydx * ctl1r / state.cone
        * np_pow(
            ((90.0 * hemi - lat) * RAD_PER_DEG / 2.0).tan()
                / ((90.0 * hemi - spec.truelat1) * RAD_PER_DEG / 2.0).tan(),
            state.cone,
        );
    let arg = state.cone * (deltalon * RAD_PER_DEG);
    let x = state.polei + hemi * rm * arg.sin();
    let y = state.polej - rm * arg.cos();
    (hemi * x, hemi * y)
}

/// Map scale factor m(lat); equals 1 at both true latitudes.
pub fn map_factor(state: &LambertState, spec: &GridSpec, hemi: f64, lat: f64) -> f64 {
    let half_colat = (90.0 * hemi - lat) * RAD_PER_DEG / 2.0;
    let half_colat1 = (90.0 * hemi - spec.truelat1) * RAD_PER_DEG / 2.0;
    (spec.truelat1 * RAD_PER_DEG).cos() / (lat * RAD_PER_DEG).cos()
        * np_pow(half_colat.tan() / half_colat1.tan(), state.cone)
}

/// (SINALPHA, COSALPHA): alpha = cone * (stand_lon - lon), geogrid
/// get_rotang PROJ_LC (no hemisphere factor).
pub fn rotation(state: &LambertState, spec: &GridSpec, lon: f64) -> (f64, f64) {
    let alpha = state.cone * RAD_PER_DEG * wrap180(spec.stand_lon - lon);
    (alpha.sin(), alpha.cos())
}
